server/core/storage.go

```diff
@@ -515,25 +553,52 @@ func (s *Storage) initServiceGCSafePointForGCWorker() (*ServiceSafePoint, error)
 
 // LoadMinServiceGCSafePoint returns the minimum safepoint across all services
 func (s *Storage) LoadMinServiceGCSafePoint(now time.Time) (*ServiceSafePoint, error) {
-	prefix := path.Join(gcPath, "safe_point", "service") + "/"
-	prefixEnd := clientv3.GetPrefixRangeEnd(prefix)
-	keys, values, err := s.LoadRange(prefix, prefixEnd, 0)
+	loadAll := func() ([]string, []*ServiceSafePoint, error) {
+		prefix := path.Join(gcPath, "safe_point", "service") + "/"
+		prefixEnd := clientv3.GetPrefixRangeEnd(prefix)
+		keys, values, err := s.LoadRange(prefix, prefixEnd, 0)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		allServiceSafePoints := make([]*ServiceSafePoint, 0, len(values))
+		for _, value := range values {
+			ssp := &ServiceSafePoint{}
+			if err := json.Unmarshal([]byte(value), ssp); err != nil {
+				return nil, nil, err
+			}
+			allServiceSafePoints = append(allServiceSafePoints, ssp)
+		}
+		return keys, allServiceSafePoints, nil
+	}
+
+	keys, allServiceSafePoints, err := loadAll()
 	if err != nil {
 		return nil, err
 	}
-	if len(keys) == 0 {
-		// There's no service safepoint. Store an initial value for GC worker.
-		return s.initServiceGCSafePointForGCWorker()
+
+	modified, err := s.fixGCWorkerServiceSafePpoint(allServiceSafePoints)
+	if err != nil {
+		return nil, err
 	}
 
-	min := &ServiceSafePoint{SafePoint: math.MaxUint64}
-	for i, key := range keys {
-		ssp := &ServiceSafePoint{}
-		if err := json.Unmarshal([]byte(values[i]), ssp); err != nil {
+	if modified {
+		// Reload the safepoints
+		keys, allServiceSafePoints, err = loadAll()
```

Comment:
What's more, it sames that `loadAll()` will not affect the result of `min`... I think we can remove this logic and `modified`.

Reply:
The user needs the minimum value among all non-expired safepoints, but I want to fix it with the minimum value including expired-but-not-deleted ones. So the final min value may decrease. It's possible to avoid reloading and do all these things in one loop, but trying to make the code more readable, I choose the less-effective way.

Reply:
fixed